x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/blobstore/cache/BlobStoreCacheService.java

```diff
@@ -0,0 +1,322 @@
+                        builder.startObject("version");
+                        builder.field("type", "integer");
+                        builder.endObject();
+                    }
+                    {
+                        builder.startObject("repository");
+                        builder.field("type", "keyword");
+                        builder.endObject();
+                    }
+                    {
+                        builder.startObject("blob");
+                        builder.field("type", "object");
+                        {
+                            builder.startObject("properties");
+                            {
+                                builder.startObject("name");
+                                builder.field("type", "keyword");
+                                builder.endObject();
+                                builder.startObject("path");
+                                builder.field("type", "keyword");
+                                builder.endObject();
+                            }
+                            builder.endObject();
+                        }
+                        builder.endObject();
+                    }
+                    {
+                        builder.startObject("data");
+                        builder.field("type", "object");
+                        {
+                            builder.startObject("properties");
+                            {
+                                builder.startObject("content");
+                                builder.field("type", "binary");
+                                builder.endObject();
+                            }
+                            {
+                                builder.startObject("length");
+                                builder.field("type", "long");
+                                builder.endObject();
+                            }
+                            {
+                                builder.startObject("from");
+                                builder.field("type", "long");
+                                builder.endObject();
+                            }
+                            {
+                                builder.startObject("to");
+                                builder.field("type", "long");
+                                builder.endObject();
+                            }
+                            builder.endObject();
+                        }
+                        builder.endObject();
+                    }
+                    builder.endObject();
+                }
+                builder.endObject();
+            }
+            builder.endObject();
+        }
+        return builder;
+    }
+
+    public CachedBlob get(String repository, String name, String path, long offset) {
+        final PlainActionFuture<CachedBlob> future = PlainActionFuture.newFuture();
+        getAsync(repository, name, path, offset, future);
+        return future.actionGet();
```

Comment:
On further reflection, this is likely only an issue if system indices utilize searchable snapshots. Cannot think of a case where that makes sense now, so we could perhaps guard against system indices using searchable snapshots (if we do not do that already)?

Reply:
We talked about this via another channel and we agreed that there could be bad situations in the case of system indices using searchable snapshots. Elasticsearch expects system indices to be writable and searchable snapshots indices are read-only by nature; this can lead to malfunctional features. Similarly system and hidden indices can be snapshotted but we want to prevent snapshot to be mounted as system indices like `.security` or `.snapshot-blob-cache`.

I opened https://github.com/elastic/elasticsearch/pull/61517 to forbid mounting system indices. 

We also agreed on improving the timeouts around gets/puts in snapshot blob cache to allow recoveries to move forward even if the snapshot blob cache index is unavailable.

Reply:
A possible alternative idea is to bypass the cache index for indices on which the GET would run on the `SYSTEM_READ` threadpool. This would let us support them as searchable snapshots, just without the recovery speed advantage of the cache; maybe we can recommend running them with replicas so as not to need the faster startup anyway.

Adding a timeout is relatively easy, the only question remains how long should it be? What do you think about `5s`?

Reply:
Bit of a wild goose chase checking this one. We could be on a `GET` thread and I suddenly worried that we also use a `GET` thread for reading from the cache index, which would still cause a deadlock:

https://github.com/elastic/elasticsearch/blob/8c51fc7e2d35d3edb301e35d4bbb23e7030b1654/server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java#L53-L54

But then I saw that we have special handling for gets against system indices:

https://github.com/elastic/elasticsearch/blob/8c51fc7e2d35d3edb301e35d4bbb23e7030b1654/server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java#L117-L126

The constructor parameter suggests that we might first dispatch to a `GET` thread, and then dispatch again to the result of `getExecutor`, but in fact we don't we handle the transport action on `SAME` and then dispatch once. So it's all ok.

I added a 5-second timeout and some logging in eb1a08ec99b.